### DISCUSSION TITLE
Issues migrating auth from Firebase -> Supabase

### DISCUSSION THREAD

**Discussion by kaspnilsson** [supabase/supabase#5418] [Questions] [answered]
Hey all,

I have a relatively normal Firebase project that I'd like to migrate onto Supabase for all of [these](https://supabase.com/blog/2021/07/28/mobbin-supabase-200000-users) reasons. I'd like to not lose all of my user data.

Having a hard time migrating my users from Firebase into Supabase. I've tried using [this script](https://github.com/liaujianjie/firebase-to-supabase-auth-migrator) 

What I'd really like to do is import all of my users from Firebase by email, associate their firebase ID with their newly generated Supabase UUID, and use this mapping to migrate my actual user data.

I've tried doing this, and I get a table where only the `id`, `email`, and `firebase_id` fields have values... But then when I go to log in with an email magic link, I get an `HTTP 500: Database error finding user` and when I try to use a social login associated with one of those emails I get silent server error `server_error: Error checking for existing users`.

Any tips on writing users with no provider to `auth.users`, or how to debug these errors?

**Comment by GaryAustin1** [comment]
Not sure that code works anymore...  Several people have been asking about creating users (like for testing and importing).
Here is a recent thread that seems to show the difficulties with just doing table inserts...
https://github.com/supabase/supabase/discussions/5248   read towards the end.

There are other threads if you search.

There was a new table "identities" added to auth 5 months or so ago it looks like:
https://github.com/supabase/gotrue/pull/208

I'm of no help, just pointing out somethings.  Seems like a script making auth api calls maybe the way to go, but I've not really looked at it that deeply.

**Reply by liaujianjie** [reply]
I'm not sure what's broken in my script but I'll be happy to look at PRs to it. :-)

**Reply by GaryAustin1** [reply]
The poster was having issues with it working and others have been having trouble signing in "copied in" users recently.  

There was a change made after your last release to add an auth.identity table which seems to also have info for users.

But I just created a user by hand with sql in the user table, did not create an entry in the identity table and it seems to work.  So it may just be mistakes being made on those trying to do the direct inserts of user data. 

I've not run your program so maybe the poster was doing something wrong with it...

**Comment by kaspnilsson** [comment, marked as answer]
Super helpful actually. Found an incredibly hacky workaround that worked perfectly in my case.

I'm using only: `email (magic link, no password), google, facebook` for auth at this point in the migration.

In my firebase app running locally, I wrote a hook like this:

```ts

  useEffect(() => {
    const fetchAndSignUp = async () => {
      // Get users from Firebase
      const users = await getUsers([
      ])

      // Sign each one up by email with some fake password, save the firebase_id to user data
      for (const { email, uid } of users) {
        const { error } = await supabase.auth.signUp(
          { email, password: 'aboguspassword' },
          { data: { firebase_id: uid } }
        )
        if (error) {
          // Run this script w/ console open to see whatever errors arise
          debugger
        } else { console.log('Migrated user ', email) }
      }
    }
    fetchAndSignUp()
  }, [])
```

This works perfectly for my case as the password is seemingly reset every time a user uses the magic link and is totally ignored when they use a social login. After doing this, it's fairly trivial to create a lookup of `{FirebaseID => SupabaseID}` and use that to migrate ownership of your data.

**Reply by GaryAustin1** [reply]
Probably want to make that password random...  Otherwise if someone looks at it they can login to all your user accounts just knowing the email if the user has not run the magic link...

**Reply by kaspnilsson** [reply]
This is _definitely_ good advice. There are only three or four active users in my case, and I don't even provide access via password, but anyone handling actually important data should probably randomize the password!
